### PR TITLE
A farewell to the empirically proven to be failed modern OOP

### LINKED ISSUE #1: Limitation on how many entities can be saved in a single schema
The current implementation of `read-schema-from-disk` assumes that you will always have one entity identifier followed by one single entity to deserialize. Hence, multiple entities of the same type, which will all be present in the same section of the serialized schema, will not be able to be read.
The solution to this problem is to change the logic of "walking in pairs" that was implemented to a proper reading of the content between identifiers.

### PR DESCRIPTION
**Pull request by EduardoLR10**
# Goal
Once upon a time, there were a @EduardoLR10 and a @MMagueta that wanted to go against their own bias towards functional programming and immutability. They tried very hard to embrace a modern approach to separation of concerns using modern, a.k.a, post-Java, object-oriented programming for RacketowerDB.
They managed to go far but eventually, reality hit, and the OOP monument crumbled with their dreams inside. Confident that it was not a matter of bias, they were enlightened by @z-silver to change the design of the application to use a combination of immutable structs, generic interfaces, and contracts. Such a set of abstractions proved to be a far better alternative in comparison, and the threefold team dedicated a significant amount of time to making an overall design change that would allow them to dream again.

# Related Issues

Close #1.

# How to Test

- Make sure you delete all data files, i.e., within the `ndf/data` folder.
- Run `nix develop && just build && ./main` and you **should not** see any errors.

# Observations for the Reviewer

I hope you understand that this is a small step for us, but one giant leap for the future of the official database project of DrN.

### CODE REVIEW

**Comment by EduardoLR10** [comment]
You can read more about how cursed this was [here](https://duing.dev/posts/ooptypeclasses/).

File: util.rkt [hunk 1]
```diff
@@ -1,27 +1,22 @@
...
-;; (define-syntax-parse-rule (create name) 
-;;   #:with name-string (datum->syntax #f #'name)
-;;   (define name-string 2))
-
 (define (fix-empty-read-bytes-lines lines)
   (define (fix-one-turn inner-lines)
     (let ((newline-flag #f))
       (foldl (lambda (line new-lines)
-               (if newline-flag
+               [if newline-flag
```
Comment (z-silver) [inline util.rkt]: Is there a reason for the square brackets here?

File: io.rkt [hunk 2]
```diff
@@ -27,97 +31,103 @@
...
+                (off-set (* row-id total-size))
+                (file-name (build-ndf-filename table-name #:data? 'data))
+                (out (open-output-file file-name #:exists 'can-update)))
+           (file-position out off-set)
+           (write-bytes converted-row out)
+           (close-output-port out)
+           (set! schema (update-row-id-table schema table-name (+ row-id 1))))]
+        [(procedura? entity)
+         (println "Don't write procedures yet")])
+      schema))
```
Comment (z-silver) [inline io.rkt:50]: Given that this is a writing function, do we really need to return this?

File: io.rkt [hunk 3]
```diff
@@ -27,97 +31,103 @@
...
+  
+  (define (write-rows-to-disk schema table-name rows)
+    (if (empty? rows)
+        schema
```
Comment (z-silver) [inline io.rkt]: Given that this is a writing function, do we really need to return this?

File: io.rkt [hunk 4]
```diff
@@ -27,97 +31,103 @@
...
+        [(table? entity)
+         (let* ((converted-row (convert-row entity row))
+                (row-id (table-row-id entity))
+                (total-size (fields-size (table-fields entity)))
+                (off-set (* row-id total-size))
+                (file-name (build-ndf-filename table-name #:data? 'data))
+                (out (open-output-file file-name #:exists 'can-update)))
+           (file-position out off-set)
+           (write-bytes converted-row out)
+           (close-output-port out)
```
Comment (z-silver) [inline io.rkt]: We should use [`call-with-output-file`](https://docs.racket-lang.org/reference/file-ports.html#%28def._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._call-with-output-file%29%29) here instead.

File: io.rkt [hunk 5]
```diff
@@ -27,97 +31,103 @@
...
-                (file-name (build-ndf-filename table-name #:data? 'data))
-                (out (open-output-file file-name #:exists 'can-update)))
-           (file-position out off-set)
-           (write-bytes converted-row out)
-           (close-output-port out)
-           (set! schema (update-row-id-table schema table-name (+ row-id 1))))]
-        [(is-a? entity procedure%)
-         (println "Don't write procedures yet")])
-      schema)))
+      (close-output-port out))))
```
Comment (z-silver) [inline io.rkt]: We should use [`call-with-output-file`](https://docs.racket-lang.org/reference/file-ports.html#%28def._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._call-with-output-file%29%29) here instead.

File: io.rkt [hunk 6]
```diff
@@ -27,97 +31,103 @@
...
          (raise 'tried-update-row-id-with-procedure)])))
 
   (define (write-table-to-disk table table-name)
-    (let* ((serialized-table (send table serialize))
+    (let* ((serialized-table (serialize table))
            (file-name (build-ndf-filename table-name))
-           (out (open-output-file file-name #:exists 'can-update)))
+           (out (open-output-file file-name #:exists 'truncate)))
       (write-bytes serialized-table out)
       (close-output-port out)))
```
Comment (z-silver) [inline io.rkt]: We should use [`call-with-output-file`](https://docs.racket-lang.org/reference/file-ports.html#%28def._%28%28lib._racket%2Fprivate%2Fbase..rkt%29._call-with-output-file%29%29) here instead.

File: ast.rkt [hunk 7]
```diff
@@ -1,135 +1,122 @@
...
+(define-serializable type
+  (name byte-size) #:transparent
+  #:methods gen:byteable
+    [(define (from-bytes self byte-stream)
       (let ((received-bytes-size (bytes-length byte-stream)))
-        (if (eq? received-bytes-size byte-size)
-            (case (list 'quote (string->symbol name))
-              [('INTEGER) (new integer32% [value (integer-bytes->integer byte-stream #t)])]
-              [('VARCHAR) (new string% [value (bytes->string/utf-8 byte-stream)])]
+        (if (eq? received-bytes-size (type-byte-size self))
```
Comment (z-silver) [inline ast.rkt]: Unless the size of types is guaranteed not to be larger than a fixnum, we should use `equal?` instead of `eq?`.

File: ast.rkt [hunk 8]
```diff
@@ -1,135 +1,122 @@
...
-      (super-new)))
+(define-serializable procedure
+  (identifier) #:transparent
+  #:methods gen:identifiable
+  [(define (give-identifier self)
+     (procedure-identifier self))]
+  #:methods gen:serializable
+  [(define (serialize _self #:size [_size #f])
+     (bytes-append (string->bytes/utf-8 "procedures' serialization is not yet implemented")))
+   {define (deserialize _self byte-stream)
```
Comment (z-silver) [inline ast.rkt]: Is there a reason for the curly braces here?

File: io.rkt [hunk 9]
```diff
@@ -27,97 +31,103 @@
...
-        (if (even? i)
-            (set! builder-class (hash-ref entity-classes (bytes->string/utf-8 (list-ref read-lines i))))
-            (set! schema (hash-union schema (make-immutable-hash
-                                             (send
-                                              (make-object builder-class)
-                                              deserialize-hash-list
-                                              (list-ref read-lines i)
-                                              '()))))))
-      (make-hash (hash->list schema))))
+           (real-lines (port->bytes-lines in))
```
Comment (z-silver) [inline io.rkt]:
```suggestion
           (real-lines (port->bytes-lines in #:close? #t))
```

File: io.rkt [hunk 10]
```diff
@@ -27,97 +31,103 @@
...
+          make-hash)))
 
   (define (read-table-from-disk table-name)
-    (let* ((file-name (build-ndf-filename table-name))
-           (in (open-input-file file-name #:mode 'binary))
-           (table (new table%)))
-      (send table deserialize (port->bytes in))
+    (let* ((file-name (build-ndf-filename #:data? 'entity table-name))
+           (in (open-input-file file-name #:mode 'binary)))
+      (define-values (table table-consumed) (deserialize struct:table (port->bytes in)))
```
Comment (z-silver) [inline io.rkt]:
```suggestion
      (define-values (table table-consumed) (deserialize struct:table (port->bytes in #:close? #t)))
```

**Review by z-silver — CHANGES_REQUESTED** [review]

File: io.rkt (hunk 2 above)
Comment (EduardoLR10) [reply io.rkt:50]: Yes, we do. A writing process of values of a table also updates the field `row-id`  which controls how many bytes we will jump as offset in order to preserve previously written data.

File: io.rkt (hunk 3 above)
Comment (EduardoLR10) [reply io.rkt]: Yes, we do. A writing process of values of a table also updates the field `row-id` which controls how many bytes we will jump as offset in order to preserve previously written data.

File: io.rkt [hunk 11]
```diff
@@ -27,23 +31,24 @@
...
+         (let* ((converted-row (convert-row entity row))
+                (row-id (table-row-id entity))
+                (total-size (fields-size (table-fields entity)))
+                (off-set (* row-id total-size))
+                (file-name (build-ndf-filename table-name #:data? 'data)))
+           (call-with-output-file file-name
+             (lambda (out)
+               (file-position out off-set)
+               (write-bytes converted-row out))
+             #:exists 'can-update)
```
Comment (z-silver) [inline io.rkt]:
```suggestion
           (call-with-output-file file-name #:exists 'can-update
             (lambda (out)
               (file-position out off-set)
               (write-bytes converted-row out)))
```

File: io.rkt [hunk 12]
```diff
@@ -52,92 +57,99 @@
...
+           schema)]
+        [(procedura? entity)
+         (raise 'tried-update-row-id-with-procedure)])))
+
+  (define (write-table-to-disk table table-name)
+    (let* ((serialized-table (serialize table))
+           (file-name (build-ndf-filename table-name)))
+      (call-with-output-file file-name
+        (curry write-bytes serialized-table)
+         #:exists 'truncate)))
```
Comment (z-silver) [inline io.rkt]:
```suggestion
      (call-with-output-file file-name #:exists 'truncate
        (curry write-bytes serialized-table))))
```

File: io.rkt [hunk 13]
```diff
@@ -52,92 +57,99 @@
...
-           (set! schema (update-row-id-table schema table-name (+ row-id 1))))]
-        [(is-a? entity procedure%)
-         (println "Don't write procedures yet")])
-      schema)))
+           (file-name (build-ndf-filename "schema" #:data? 'schema)))
+      (call-with-output-file file-name
+        (lambda (out)
+          (~>> (group-by (compose give-identifier cdr) schema-list)
+               (map (curry write-entity-to-disk out))))
+        #:exists 'truncate))))
```
Comment (z-silver) [inline io.rkt]:
```suggestion
      (call-with-output-file file-name #:exists 'truncate
        (lambda (out)
          (~>> (group-by (compose give-identifier cdr) schema-list)
               (map (curry write-entity-to-disk out))))))))
```

**Review by z-silver — CHANGES_REQUESTED** [review]
Just some minor readability improvements this time.

**Review by z-silver — APPROVED** [review]

**Pull request merged**